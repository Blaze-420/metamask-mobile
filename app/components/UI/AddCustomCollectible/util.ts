import { isValidAddress } from 'ethereumjs-util';
import { isSmartContractAddress } from '../../../util/transactions';
import { strings } from '../../../../locales/i18n';

interface ValidationResult {
  isValid: boolean;
  warningMessage: string;
}

/**
 * Validates a custom collectible address
 * @param address - The address to validate
 * @param chainId - The chain ID to check if the address is a smart contract
 * @returns Promise<ValidationResult> - Object containing validation result and warning message
 */
export const validateCustomCollectibleAddress = async (
  address: string,
  chainId?: string | null,
): Promise<ValidationResult> => {
  const isValidEthAddress = isValidAddress(address);

  if (address.length === 0) {
    return {
      isValid: false,
      warningMessage: strings('collectible.address_cant_be_empty'),
    };
  }

  if (!isValidEthAddress) {
    return {
      isValid: false,
      warningMessage: strings('collectible.address_must_be_valid'),
    };
  }

  if (chainId && !(await isSmartContractAddress(address, chainId))) {
    return {
      isValid: false,
      warningMessage: strings('collectible.address_must_be_smart_contract'),
    };
  }

  return {
    isValid: true,
    warningMessage: '',
  };
};

interface OwnershipValidationResult {
  isOwner: boolean;
  error?: {
    title: string;
    message: string;
  };
}

/**
 * Validates if a user owns a specific NFT
 * @param selectedAddress - The user's address
 * @param contractAddress - The NFT contract address
 * @param tokenId - The NFT token ID
 * @returns Promise<OwnershipValidationResult> - Object containing ownership result and optional error
 */
export const validateCollectibleOwnership = async (
  selectedAddress: string,
  contractAddress: string,
  tokenId: string,
): Promise<OwnershipValidationResult> => {
  try {
    const { NftController } = Engine.context;
    const isOwner = await NftController.isNftOwner(
      selectedAddress,
      contractAddress,
      tokenId,
    );

    if (!isOwner) {
      return {
        isOwner: false,
        error: {
          title: strings('collectible.not_owner_error_title'),
          message: strings('collectible.not_owner_error'),
        },
      };
    }

    return { isOwner: true };
  } catch {
    return {
      isOwner: false,
      error: {
        title: strings('collectible.ownership_verification_error_title'),
        message: strings('collectible.ownership_verification_error'),
      },
    };
  }
};
