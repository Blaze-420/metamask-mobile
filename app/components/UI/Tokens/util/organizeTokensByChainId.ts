import { TokenI } from '../types';

/**
 * Organizes an array of tokens into groups by chainId
 * @param tokens Array of tokens to organize
 * @returns Object with chainId keys and arrays of tokens as values
 */
export const organizeTokensByChainId = (
  tokens: TokenI[],
): { [chainId: string]: TokenI[] } =>
  tokens.reduce<{ [chainId: string]: TokenI[] }>((acc, token) => {
    if (!token.chainId) {
      return acc;
    }

    if (!acc[token.chainId]) {
      acc[token.chainId] = [];
    }

    acc[token.chainId].push(token);

    return acc;
  }, {});
