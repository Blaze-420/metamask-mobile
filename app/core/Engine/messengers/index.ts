import { getAccountsControllerMessenger } from './accounts-controller-messenger';
import type { ControllerMessengerByControllerName } from '../types';
import { getEarnControllerMessenger } from './earn-controller-messenger';

/**
 * The messengers for the controllers that have been.
 */
export const CONTROLLER_MESSENGERS: ControllerMessengerByControllerName = {
  AccountsController: {
    getMessenger: getAccountsControllerMessenger,
  },
  EarnController: {
    getMessenger: getEarnControllerMessenger,
  },
} as const;
