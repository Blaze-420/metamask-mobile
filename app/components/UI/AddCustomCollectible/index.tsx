import React, { useState, useEffect, useCallback } from 'react';
import { useSelector } from 'react-redux';
import { Alert, Text, TextInput, View, StyleSheet } from 'react-native';
import { fontStyles } from '../../../styles/common';
import Engine from '../../../core/Engine';
import { strings } from '../../../../locales/i18n';
import ActionView from '../ActionView';
import Device from '../../../util/device';
import { MetaMetricsEvents } from '../../../core/Analytics';
import {
  validateCustomCollectibleAddress as validateAddress,
  validateCollectibleOwnership as validateOwnership,
  validateCustomCollectibleTokenId as validateTokenId,
} from './util';
import { useTheme } from '../../../util/theme';
import { NFTImportScreenSelectorsIDs } from '../../../../e2e/selectors/wallet/ImportNFTView.selectors';
import { selectSelectedInternalAccountFormattedAddress } from '../../../selectors/accountsController';
import { getDecimalChainId } from '../../../util/networks';
import { useMetrics } from '../../../components/hooks/useMetrics';
import Logger from '../../../util/Logger';
import { NetworkSelectorDropdown } from '../AddCustomToken/NetworkSelectorDropdown';
import { Hex } from '@metamask/utils';
import { TraceName, endTrace, trace } from '../../../util/trace';

// TODO: Replace "any" with type
// eslint-disable-next-line @typescript-eslint/no-explicit-any
const createStyles = (colors: any) =>
  StyleSheet.create({
    wrapper: {
      backgroundColor: colors.background.default,
      flex: 1,
    },
    rowWrapper: {
      padding: 20,
    },
    rowTitleText: {
      paddingBottom: 3,
      // TODO: Replace "any" with type
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      ...(fontStyles.normal as any),
      color: colors.text.default,
    },
    textInput: {
      borderWidth: 1,
      borderRadius: 4,
      borderColor: colors.border.default,
      padding: 16,
      // TODO: Replace "any" with type
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      ...(fontStyles.normal as any),
      color: colors.text.default,
    },
    warningText: {
      marginTop: 15,
      color: colors.error.default,
      // TODO: Replace "any" with type
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      ...(fontStyles.normal as any),
    },
  });

interface AddCustomCollectibleProps {
  // TODO: Replace "any" with type
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  navigation?: any;
  collectibleContract?: {
    address: string;
  };
  setOpenNetworkSelector: (val: boolean) => void;
  selectedNetwork: string | null;
  chainId: string | null;
}

const AddCustomCollectible = ({
  navigation,
  collectibleContract,
  setOpenNetworkSelector,
  selectedNetwork,
  chainId,
}: AddCustomCollectibleProps) => {
  const [mounted, setMounted] = useState<boolean>(true);
  const [address, setAddress] = useState<string>('');
  const [tokenId, setTokenId] = useState<string>('');
  const [warningAddress, setWarningAddress] = useState<string>('');
  const [warningTokenId, setWarningTokenId] = useState<string>('');
  const [inputWidth, setInputWidth] = useState<string | undefined>(
    Device.isAndroid() ? '99%' : undefined,
  );
  const [loading, setLoading] = useState(false);
  // TODO: Replace "any" with type
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const assetTokenIdInput = React.createRef() as any;
  const { colors, themeAppearance } = useTheme();
  const { trackEvent, createEventBuilder } = useMetrics();
  const styles = createStyles(colors);

  const selectedAddress = useSelector(
    selectSelectedInternalAccountFormattedAddress,
  );

  useEffect(() => {
    setMounted(true);
    // Workaround https://github.com/facebook/react-native/issues/9958
    inputWidth &&
      setTimeout(() => {
        mounted && setInputWidth('100%');
      }, 100);
    collectibleContract && setAddress(collectibleContract.address);
    return () => {
      setMounted(false);
    };
  }, [mounted, collectibleContract, inputWidth]);

  const getAnalyticsParams = useCallback(() => {
    try {
      return {
        chain_id: getDecimalChainId(chainId),
        source: 'manual',
      };
    } catch (error) {
      Logger.error(error as Error, 'AddCustomCollectible.getAnalyticsParams');
      return undefined;
    }
  }, [chainId]);

  const validateCustomCollectibleAddress = async (): Promise<boolean> => {
    const result = await validateAddress(address, chainId);
    setWarningAddress(result.warningMessage);
    return result.isValid;
  };

  const validateCustomCollectibleTokenId = (): boolean => {
    const result = validateTokenId(tokenId);
    setWarningTokenId(result.warningMessage);
    return result.isValid;
  };

  const validateCustomCollectible = async (): Promise<boolean> => {
    const validatedAddress = await validateCustomCollectibleAddress();
    const validatedTokenId = validateCustomCollectibleTokenId();
    return validatedAddress && validatedTokenId;
  };

  const validateCollectibleOwnership = async (): Promise<boolean> => {
    if (!address) return false;
    const result = await validateOwnership(
      selectedAddress ?? '',
      address,
      tokenId,
    );
    if (!result.isOwner && result.error) {
      Alert.alert(result.error.title, result.error.message);
    }
    return result.isOwner;
  };

  const addNft = async (): Promise<void> => {
    setLoading(true);
    if (!(await validateCustomCollectible())) {
      setLoading(false);
      return;
    }
    if (!(await validateCollectibleOwnership())) {
      setLoading(false);
      return;
    }

    // TODO: Replace "any" with type
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    const { NftController } = Engine.context as any;
    NftController.addNft(address, tokenId, chainId);

    trace({ name: TraceName.ImportNfts });

    await NftController.addNft(address, tokenId);

    endTrace({ name: TraceName.ImportNfts });

    const params = getAnalyticsParams();
    if (params) {
      trackEvent(
        createEventBuilder(MetaMetricsEvents.COLLECTIBLE_ADDED)
          .addProperties(params)
          .build(),
      );
    }

    setLoading(false);
    navigation.goBack();
  };

  const cancelAddCollectible = (): void => {
    navigation.goBack();
  };

  const onAddressChange = (newAddress: string): void => {
    setAddress(newAddress);
  };

  const onTokenIdChange = (newTokenId: string): void => {
    setTokenId(newTokenId);
  };

  const jumpToAssetTokenId = (): void => {
    assetTokenIdInput.current?.focus();
  };

  return (
    <View style={styles.wrapper} testID={NFTImportScreenSelectorsIDs.CONTAINER}>
      <ActionView
        cancelText={strings('add_asset.collectibles.cancel_add_collectible')}
        confirmText={strings('add_asset.collectibles.add_collectible')}
        onCancelPress={cancelAddCollectible}
        onConfirmPress={addNft}
        confirmDisabled={!address || !tokenId || !selectedNetwork}
        loading={loading}
      >
        <View>
          <View style={styles.rowWrapper}>
            <NetworkSelectorDropdown
              setOpenNetworkSelector={setOpenNetworkSelector}
              selectedNetwork={selectedNetwork ?? ''}
              chainId={chainId as Hex}
            />
          </View>
          <View style={styles.rowWrapper}>
            <Text style={styles.rowTitleText}>
              {strings('collectible.collectible_address')}
            </Text>
            <TextInput
              style={[
                styles.textInput,
                inputWidth ? { width: inputWidth } : {},
              ]}
              placeholder={'0x...'}
              placeholderTextColor={colors.text.muted}
              value={address}
              onChangeText={onAddressChange}
              onBlur={validateCustomCollectibleAddress}
              testID={NFTImportScreenSelectorsIDs.ADDRESS_INPUT_BOX}
              onSubmitEditing={jumpToAssetTokenId}
              keyboardAppearance={themeAppearance}
            />
            <Text
              style={styles.warningText}
              testID={NFTImportScreenSelectorsIDs.ADDRESS_WARNING_MESSAGE}
            >
              {warningAddress}
            </Text>
          </View>
          <View style={styles.rowWrapper}>
            <Text style={styles.rowTitleText}>
              {strings('collectible.collectible_token_id')}
            </Text>
            <TextInput
              style={[
                styles.textInput,
                inputWidth ? { width: inputWidth } : {},
              ]}
              value={tokenId}
              keyboardType="numeric"
              onChangeText={onTokenIdChange}
              onBlur={validateCustomCollectibleTokenId}
              testID={NFTImportScreenSelectorsIDs.IDENTIFIER_INPUT_BOX}
              ref={assetTokenIdInput}
              onSubmitEditing={addNft}
              returnKeyType={'done'}
              placeholder={strings('collectible.id_placeholder')}
              placeholderTextColor={colors.text.muted}
              keyboardAppearance={themeAppearance}
            />
            <Text
              style={styles.warningText}
              testID={NFTImportScreenSelectorsIDs.IDENTIFIER_WARNING_MESSAGE}
            >
              {warningTokenId}
            </Text>
          </View>
        </View>
      </ActionView>
    </View>
  );
};

export default AddCustomCollectible;
