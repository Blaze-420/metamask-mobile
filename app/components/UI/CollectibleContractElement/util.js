import Engine from '../../../core/Engine';
import { toHex } from '@metamask/controller-utils';

export const removeNft = ({
  selectedAddress,
  chainId,
  longPressedCollectible,
  removeFavoriteCollectible,
  networkConfigurations,
}) => {
  const { NftController } = Engine.context;
  removeFavoriteCollectible(
    selectedAddress,
    chainId,
    longPressedCollectible.current,
  );
  const chainIdHex = toHex(longPressedCollectible.current.chainId);
  const config = networkConfigurations[chainIdHex];
  const nftNetworkClientId =
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-ignore
    config?.rpcEndpoints?.[config.defaultRpcEndpointIndex]?.networkClientId;

  NftController.removeAndIgnoreNft(
    longPressedCollectible.current.address,
    longPressedCollectible.current.tokenId,
    {
      networkClientId: nftNetworkClientId,
      userAddress: selectedAddress.toLowerCase(),
    },
  );
};
