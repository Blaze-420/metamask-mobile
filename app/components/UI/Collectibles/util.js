import Engine from '../../../core/Engine';
import { toHex } from '@metamask/controller-utils';
import { strings } from '../../../../locales/i18n';
import { Alert } from 'react-native';

export const refreshMetadata = (
  collectible,
  networkConfigurations,
  userAddress,
) => {
  const { NftController } = Engine.context;

  const chainIdHex = toHex(collectible.chainId);
  const config = networkConfigurations[chainIdHex];
  const nftNetworkClientId =
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-ignore
    config?.rpcEndpoints?.[config.defaultRpcEndpointIndex]?.networkClientId;

  NftController.addNft(collectible.address, collectible.tokenId, {
    networkClientId: nftNetworkClientId,
    userAddress,
  });
};

export const removeNft = (
  collectible,
  networkConfigurations,
  selectedAddress,
) => {
  const { NftController } = Engine.context;

  const chainIdHex = toHex(collectible.chainId);
  const config = networkConfigurations[chainIdHex];
  const nftNetworkClientId =
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-ignore
    config?.rpcEndpoints?.[config.defaultRpcEndpointIndex]?.networkClientId;

  NftController.removeAndIgnoreNft(collectible.address, collectible.tokenId, {
    networkClientId: nftNetworkClientId,
    userAddress: selectedAddress.toLowerCase(),
  });
  Alert.alert(
    strings('wallet.collectible_removed_title'),
    strings('wallet.collectible_removed_desc'),
  );
};
