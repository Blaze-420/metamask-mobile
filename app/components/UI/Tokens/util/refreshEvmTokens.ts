import { Hex } from '@metamask/utils';
import Engine from '../../../../core/Engine';
import Logger from '../../../../util/Logger';

interface RefreshEvmTokensProps {
  isEvmSelected: boolean;
  evmNetworkConfigurationsByChainId: Record<
    string,
    { chainId: Hex; nativeCurrency: string }
  >;
  nativeCurrencies: string[];
}

export const refreshEvmTokens = async ({
  isEvmSelected,
  evmNetworkConfigurationsByChainId,
  nativeCurrencies,
}: RefreshEvmTokensProps) => {
  if (!isEvmSelected) {
    return;
  }

  const {
    TokenDetectionController,
    AccountTrackerController,
    CurrencyRateController,
    TokenRatesController,
    TokenBalancesController,
    NetworkController,
  } = Engine.context;

  const networkClientIds = Object.values(
    NetworkController.state.networkConfigurationsByChainId,
  ).map(
    (network) =>
      network?.rpcEndpoints?.[network.defaultRpcEndpointIndex]?.networkClientId,
  );

  const actions = [
    TokenDetectionController.detectTokens({
      chainIds: Object.keys(evmNetworkConfigurationsByChainId) as Hex[],
    }),
    TokenBalancesController.updateBalances({
      chainIds: Object.keys(evmNetworkConfigurationsByChainId) as Hex[],
    }),

    AccountTrackerController.refresh(networkClientIds),
    CurrencyRateController.updateExchangeRate(nativeCurrencies),
    ...Object.values(evmNetworkConfigurationsByChainId).map((network) =>
      TokenRatesController.updateExchangeRatesByChainId([
        {
          chainId: network.chainId,
          nativeCurrency: network.nativeCurrency,
        },
      ]),
    ),
  ];

  await Promise.all(actions).catch((error) => {
    Logger.error(error, 'Error while refreshing tokens');
  });
};
