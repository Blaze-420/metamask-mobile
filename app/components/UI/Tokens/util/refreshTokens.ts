import { Hex } from '@metamask/utils';
import Engine from '../../../../core/Engine';
import Logger from '../../../../util/Logger';
import { InternalAccount } from '@metamask/keyring-internal-api';

interface RefreshTokensProps {
  isEvmSelected: boolean;
  evmNetworkConfigurationsByChainId: Record<
    string,
    { chainId: Hex; nativeCurrency: string }
  >;
  nativeCurrencies: string[];
  selectedAccount?: InternalAccount;
}

export const refreshTokens = async ({
  isEvmSelected,
  evmNetworkConfigurationsByChainId,
  nativeCurrencies,
  selectedAccount,
}: RefreshTokensProps) => {
  if (!isEvmSelected) {
    const { MultichainBalancesController } = Engine.context;
    if (selectedAccount) {
      const actions = [
        MultichainBalancesController.updateBalance(selectedAccount?.id),
      ];
      await Promise.all(actions).catch((error) => {
        Logger.error(error, 'Error while refreshing NonEvm tokens');
      });
    }
  } else {
    const {
      TokenDetectionController,
      AccountTrackerController,
      CurrencyRateController,
      TokenRatesController,
      TokenBalancesController,
      NetworkController,
    } = Engine.context;

    const networkClientIds = Object.values(
      NetworkController.state.networkConfigurationsByChainId,
    ).map(
      (network) =>
        network?.rpcEndpoints?.[network.defaultRpcEndpointIndex]
          ?.networkClientId,
    );

    const actions = [
      TokenDetectionController.detectTokens({
        chainIds: Object.keys(evmNetworkConfigurationsByChainId) as Hex[],
      }),
      TokenBalancesController.updateBalances({
        chainIds: Object.keys(evmNetworkConfigurationsByChainId) as Hex[],
      }),
      AccountTrackerController.refresh(networkClientIds),
      CurrencyRateController.updateExchangeRate(nativeCurrencies),
      ...Object.values(evmNetworkConfigurationsByChainId).map((network) =>
        TokenRatesController.updateExchangeRatesByChainId([
          {
            chainId: network.chainId,
            nativeCurrency: network.nativeCurrency,
          },
        ]),
      ),
    ];

    await Promise.all(actions).catch((error) => {
      Logger.error(error, 'Error while refreshing tokens');
    });
  }
};
