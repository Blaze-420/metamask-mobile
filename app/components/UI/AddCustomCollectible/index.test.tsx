import React from 'react';
import { shallow } from 'enzyme';
import AddCustomCollectible from './';
import configureMockStore from 'redux-mock-store';
import { Provider } from 'react-redux';
import initialRootState from '../../../util/test/initial-root-state';
import {
  validateCustomCollectibleAddress,
  validateCollectibleOwnership,
} from './util';
import { isValidAddress } from 'ethereumjs-util';
import { isSmartContractAddress } from '../../../util/transactions';
import Engine from '../../../core/Engine';

const mockStore = configureMockStore();

const store = mockStore(initialRootState);

jest.mock('react-redux', () => ({
  ...jest.requireActual('react-redux'),
  useSelector: jest.fn().mockImplementation(() => ''),
}));

jest.mock('ethereumjs-util', () => ({
  isValidAddress: jest.fn(),
}));

jest.mock('../../../util/transactions', () => ({
  isSmartContractAddress: jest.fn(),
}));

jest.mock('../../../../locales/i18n', () => ({
  strings: jest.fn((key) => key),
}));

jest.mock('../../../core/Engine', () => ({
  context: {
    NftController: {
      isNftOwner: jest.fn(),
    },
  },
}));

describe('AddCustomCollectible', () => {
  it('should render correctly', () => {
    const wrapper = shallow(
      <Provider store={store}>
        <AddCustomCollectible
          setOpenNetworkSelector={jest.fn()}
          selectedNetwork={'mainnet'}
          chainId={'0x1'}
        />
      </Provider>,
    );
    expect(wrapper).toMatchSnapshot();
  });
});

describe('validateCustomCollectibleAddress', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('should return invalid when address is empty', async () => {
    const result = await validateCustomCollectibleAddress('', '0x1');
    expect(result).toEqual({
      isValid: false,
      warningMessage: 'collectible.address_cant_be_empty',
    });
  });

  it('should return invalid when address is not a valid Ethereum address', async () => {
    (isValidAddress as jest.Mock).mockReturnValue(false);
    const result = await validateCustomCollectibleAddress(
      'invalid-address',
      '0x1',
    );
    expect(result).toEqual({
      isValid: false,
      warningMessage: 'collectible.address_must_be_valid',
    });
  });

  it('should return invalid when address is not a smart contract', async () => {
    (isValidAddress as jest.Mock).mockReturnValue(true);
    (isSmartContractAddress as jest.Mock).mockResolvedValue(false);
    const result = await validateCustomCollectibleAddress('0x123', '0x1');
    expect(result).toEqual({
      isValid: false,
      warningMessage: 'collectible.address_must_be_smart_contract',
    });
  });

  it('should return valid when address is a valid smart contract', async () => {
    (isValidAddress as jest.Mock).mockReturnValue(true);
    (isSmartContractAddress as jest.Mock).mockResolvedValue(true);
    const result = await validateCustomCollectibleAddress('0x123', '0x1');
    expect(result).toEqual({
      isValid: true,
      warningMessage: '',
    });
  });

  it('should not check smart contract when chainId is not provided', async () => {
    (isValidAddress as jest.Mock).mockReturnValue(true);
    const result = await validateCustomCollectibleAddress('0x123');
    expect(result).toEqual({
      isValid: true,
      warningMessage: '',
    });
    expect(isSmartContractAddress).not.toHaveBeenCalled();
  });
});

describe('validateCollectibleOwnership', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('should return false with error when verification fails', async () => {
    (Engine.context.NftController.isNftOwner as jest.Mock).mockRejectedValue(
      new Error('Verification failed'),
    );
    const result = await validateCollectibleOwnership('0x123', '0x456', '789');
    expect(result).toEqual({
      isOwner: false,
      error: {
        title: 'collectible.ownership_verification_error_title',
        message: 'collectible.ownership_verification_error',
      },
    });
  });
});
