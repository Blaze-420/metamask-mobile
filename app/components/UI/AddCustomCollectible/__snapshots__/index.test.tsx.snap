// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddCustomCollectible should render correctly 1`] = `
<ContextProvider
  value={
    {
      "getServerState": undefined,
      "noopCheck": "once",
      "stabilityCheck": "once",
      "store": {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "subscription": {
        "addNestedSub": [Function],
        "getListeners": [Function],
        "handleChangeWrapper": [Function],
        "isSubscribed": [Function],
        "notifyNestedSubs": [Function],
        "trySubscribe": [Function],
        "tryUnsubscribe": [Function],
      },
    }
  }
>
  <AddCustomCollectible
    chainId="0x1"
    selectedNetwork={
      {
        "blockExplorerUrls": [
          "https://etherscan.io",
        ],
        "chainId": "0x1",
        "defaultBlockExplorerUrlIndex": 0,
        "defaultRpcEndpointIndex": 0,
        "name": "mainnet",
        "nativeCurrency": "ETH",
        "rpcEndpoints": [
          {
            "networkClientId": "mainnet",
            "type": "infura",
            "url": "https://mainnet.infura.io/v3/{infuraProjectId}",
          },
        ],
      }
    }
    setOpenNetworkSelector={[MockFunction]}
  />
</ContextProvider>
`;
